src/main/scala/tech/sourced/gemini/FeaturesHash.scala

```diff
@@ -0,0 +1,62 @@
+package tech.sourced.gemini
+
+import tech.sourced.featurext.generated.service.Feature
+
+import scala.collection.mutable
+
+object FeaturesHash {
+  // All params below must be the same for hash & query
+  val defaultSeed = 0
+  val defaultSampleSize = 128
+  // Depend on threshold and sample_size
+  // this values calculated for threshold=0.8 and sample_size=128
+  val defaultHashtablesNum = 9
+  val defaultBandSize = 13
+
+  private var tokensSize: Int = _
+  private var wmh: WeightedMinHash = _
+
+  def hashFeatures(
+                    docFreq: OrderedDocFreq,
+                    features: Iterable[Feature],
+                    sampleSize: Int = defaultSampleSize,
+                    seed: Int = defaultSeed): Array[Array[Long]] = {
+    val OrderedDocFreq(docs, tokens, df) = docFreq
+
+    if (wmh == null || tokensSize != tokens.size) {
+      wmh = new WeightedMinHash(tokens.size, sampleSize, seed)
+    }
+
+    val bag = mutable.ArrayBuffer.fill(tokens.size)(0.toDouble)
```

Comment:
What's expected to be bigger, `features` or `tokens`?

Reply:
tokens (vocabulary of all features) in most cases will be much-much-much bigger than features (for a single file)

Reply:
Ok, so if `tokens` is an `Array[String]` and it is guaranteed to be sorted, you can improve performance by replacing the `indexOf` call with something like `java.util.Arrays.binarySearch(a.asInstanceOf[Array[AnyRef]],key)`.